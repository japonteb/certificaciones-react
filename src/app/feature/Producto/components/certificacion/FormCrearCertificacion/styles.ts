import styled from 'styled-components';

export const SpanError = styled.span`
  color: #f62d2d;
  font-weight: bold;
  font-size: 8px;
`;

export const Form = styled.form`
  width: 100%;
  margin: 2em 0;
`;

export const FormInputDiv = styled.div`
  display: flex;
  flex-direction: column;
  margin-bottom: 1em;
`;
