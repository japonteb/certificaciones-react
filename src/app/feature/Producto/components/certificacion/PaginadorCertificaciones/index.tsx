import * as PropTypes from 'prop-types';
import * as React from 'react';

interface PaginadorCertificacionesProps {
  cantidadTotalCertificaciones: number;
  onClickCambiarPagina: (numeroPagina: number) => void;
}

const CERTIFICACIONES_VISIBLES_POR_PAGINA = 10;

export const PaginadorCertificaciones: React.FC<
  PaginadorCertificacionesProps
> = ({ onClickCambiarPagina, cantidadTotalCertificaciones }) => {
  if (cantidadTotalCertificaciones <= CERTIFICACIONES_VISIBLES_POR_PAGINA) {
    return null;
  }

  const rango = Array.from(
    Array(
      Math.ceil(
        cantidadTotalCertificaciones / CERTIFICACIONES_VISIBLES_POR_PAGINA
      )
    ).keys()
  );

  return (
    <nav>
      {rango.map((index) => {
        return (
          <button
            onClick={() => onClickCambiarPagina(index)}
            key={index.toString()}
          >
            {index + 1}
          </button>
        );
      })}
    </nav>
  );
};

PaginadorCertificaciones.propTypes = {
  cantidadTotalCertificaciones: PropTypes.number.isRequired,
  onClickCambiarPagina: PropTypes.func.isRequired,
};
