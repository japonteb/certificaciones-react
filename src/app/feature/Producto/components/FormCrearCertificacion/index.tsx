import * as PropTypes from 'prop-types';
import * as React from 'react';
import * as Yup from 'yup';
import { Button } from 'app/shared/components/Button';
import { Certificacion } from '../../models/Certificacion';
import { FormikHelpers } from 'formik/dist/types';
import { Input } from 'app/shared/components/Input';
import { SpanError } from './styles';
import { useFormik } from 'formik';

interface FormValues {
  nombre: string;
  detalle: string;
  duracion: number;
  precio: number;
}

interface FormCrearCertificacionProp {
  onSubmit: (payload: Certificacion) => any;
  disabled?: boolean;
  formTitle: string;
  initialValues?: FormValues;
}

const validationSchema = Yup.object().shape<FormValues>({
  nombre: Yup.string().required('El campo nombre es requerido.'),
  detalle: Yup.string().required('El campo detalle es requerido.'),
  duracion: Yup.number()
    .required('El campo duración es requerido.')
    .typeError('La duración debe ser un número')
    .positive('El número debe ser positivo'),
  precio: Yup.number().required('El campo precio es requerido.'),
});

export const FormCrearCertificacion: React.FC<FormCrearCertificacionProp> = ({
  onSubmit,
  disabled,
  formTitle,
  initialValues = {
    nombre: '',
    detalle: '',
    duracion: 0,
    precio: 0,
  },
}) => {
  const handleSubmit = (
    values: FormValues,
    { resetForm }: FormikHelpers<FormValues>
  ) => {
    onSubmit({
      id: -1,
      nombre: values.nombre,
      detalle: values.detalle,
      duracion: values.duracion,
      precio: values.precio,
    });
    resetForm();
  };
  const formik = useFormik({
    initialValues,
    validationSchema,
    onSubmit: handleSubmit,
  });

  return (
    <form onSubmit={formik.handleSubmit}>
      <h2>{formTitle}</h2>
      <Input
        disabled={disabled}
        name="nombre"
        placeholder="Nombre de la certificación"
        value={formik.values.nombre}
        onChange={formik.handleChange}
      />
      {formik.touched.nombre && formik.errors.nombre && (
        <SpanError>{formik.errors.nombre}</SpanError>
      )}
      <Input
        disabled={disabled}
        name="detalle"
        placeholder="Detalle de la certificación"
        value={formik.values.detalle}
        onChange={formik.handleChange}
      />
      {formik.touched.detalle && formik.errors.detalle && (
        <SpanError>{formik.errors.detalle}</SpanError>
      )}
      <Input
        disabled={disabled}
        name="duracion"
        placeholder="Duración en horas"
        value={formik.values.duracion}
        onChange={formik.handleChange}
      />
      {formik.touched.duracion && formik.errors.duracion && (
        <SpanError>{formik.errors.duracion}</SpanError>
      )}

      <Input
        disabled={disabled}
        name="precio"
        placeholder="Precio en COP"
        value={formik.values.precio}
        onChange={formik.handleChange}
      />
      {formik.touched.precio && formik.errors.precio && (
        <SpanError>{formik.errors.precio}</SpanError>
      )}

      <Button type="submit">Registrar</Button>
    </form>
  );
};

FormCrearCertificacion.propTypes = {
  onSubmit: PropTypes.func.isRequired,
  formTitle: PropTypes.string.isRequired,
  disabled: PropTypes.bool,
  initialValues: PropTypes.shape({
    nombre: PropTypes.string.isRequired,
    detalle: PropTypes.string.isRequired,
    duracion: PropTypes.number.isRequired,
    precio: PropTypes.number.isRequired,
  }),
};
