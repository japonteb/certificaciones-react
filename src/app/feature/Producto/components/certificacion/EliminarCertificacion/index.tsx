import * as PropTypes from 'prop-types';
import * as React from 'react';
import { Button } from 'app/shared/components/Button';
import { Certificacion } from 'app/feature/Producto/models/Certificacion';

interface BtnEliminarCertificacionProps {
  onEliminar: (certificacion: Certificacion) => any;
  certificacion: Certificacion;
}

export const BtnEliminarCertificacion: React.FC<
  BtnEliminarCertificacionProps
> = ({ onEliminar, certificacion }) => {
  const handleEliminar = () => onEliminar(certificacion);
  return (
    <Button onClick={handleEliminar}>
      <span role="img" aria-labelledby="trash">
        üóëÔ∏è
      </span>
    </Button>
  );
};

BtnEliminarCertificacion.propTypes = {
  certificacion: PropTypes.shape({
    id: PropTypes.number.isRequired,
    nombre: PropTypes.string.isRequired,
    detalle: PropTypes.string.isRequired,
    duracion: PropTypes.number.isRequired,
    precio: PropTypes.number.isRequired,
  }).isRequired,
  onEliminar: PropTypes.func.isRequired,
};
