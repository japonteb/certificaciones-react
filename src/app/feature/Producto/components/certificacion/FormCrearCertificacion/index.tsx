import * as PropTypes from 'prop-types';
import * as React from 'react';
import * as Yup from 'yup';
import { Form, FormInputDiv, SpanError } from './styles';
import { Button } from 'app/shared/components/Button';
import { Certificacion } from 'app/feature/Producto/models/Certificacion';
import { FormikHelpers } from 'formik/dist/types';
import { Input } from 'app/shared/components/Input';
import { Label } from 'app/shared/components/Label';
import { useFormik } from 'formik';

interface FormValues {
  nombre: string;
  detalle: string;
  duracion: number;
  precio: number;
}

interface FormCrearCertificacionProp {
  onSubmit: (payload: Certificacion) => any;
  disabled?: boolean;
  formTitle: string;
  initialValues?: FormValues;
}

const validationSchema = Yup.object().shape<FormValues>({
  nombre: Yup.string().required('El campo nombre es requerido.'),
  detalle: Yup.string().required('El campo detalle es requerido.'),
  duracion: Yup.number()
    .required('El campo duración es requerido.')
    .typeError('La duración debe ser un número')
    .positive('La duración debe ser un número positivo'),
  precio: Yup.number()
    .required('El campo precio es requerido.')
    .typeError('El precio debe ser un número')
    .positive('El precio debe ser un número positivo'),
});

export const FormCrearCertificacion: React.FC<FormCrearCertificacionProp> = ({
  onSubmit,
  disabled,
  formTitle,
  initialValues = {
    nombre: '',
    detalle: '',
    duracion: 0,
    precio: 0,
  },
}) => {
  const handleSubmit = (
    values: FormValues,
    { resetForm }: FormikHelpers<FormValues>
  ) => {
    onSubmit({
      id: -1,
      nombre: values.nombre,
      detalle: values.detalle,
      duracion: Number(values.duracion),
      precio: Number(values.precio),
    });
    resetForm();
  };
  const formik = useFormik({
    initialValues,
    validationSchema,
    onSubmit: handleSubmit,
  });

  return (
    <Form onSubmit={formik.handleSubmit}>
      <h2>{formTitle}</h2>
      <FormInputDiv>
        <Label htmlFor="nombre">Nombre de la certificación:</Label>
        <Input
          disabled={disabled}
          id="nombre"
          placeholder="Ingresa el nombre"
          {...formik.getFieldProps('nombre')}
        />
        {formik.touched.nombre && formik.errors.nombre && (
          <SpanError>{formik.errors.nombre}</SpanError>
        )}
      </FormInputDiv>

      <FormInputDiv>
        <Label htmlFor="detalle">Detalle de la certificación:</Label>
        <Input
          disabled={disabled}
          id="detalle"
          placeholder="Ingresa el detalle"
          {...formik.getFieldProps('detalle')}
        />
        {formik.touched.detalle && formik.errors.detalle && (
          <SpanError>{formik.errors.detalle}</SpanError>
        )}
      </FormInputDiv>
      <FormInputDiv>
        <Label htmlFor="duracion">Duración de la certificación:</Label>
        <Input
          disabled={disabled}
          id="duracion"
          placeholder="Ingresa la duración en horas"
          {...formik.getFieldProps('duracion')}
        />
        {formik.touched.duracion && formik.errors.duracion && (
          <SpanError>{formik.errors.duracion}</SpanError>
        )}
      </FormInputDiv>
      <FormInputDiv>
        <Label htmlFor="precio">Precio de la certificación:</Label>
        <Input
          disabled={disabled}
          id="precio"
          placeholder="Ingresa el precio en COP"
          {...formik.getFieldProps('precio')}
        />
        {formik.touched.precio && formik.errors.precio && (
          <SpanError>{formik.errors.precio}</SpanError>
        )}
      </FormInputDiv>
      <Button type="submit">Registrar</Button>
    </Form>
  );
};

FormCrearCertificacion.propTypes = {
  onSubmit: PropTypes.func.isRequired,
  formTitle: PropTypes.string.isRequired,
  disabled: PropTypes.bool,
  initialValues: PropTypes.shape({
    nombre: PropTypes.string.isRequired,
    detalle: PropTypes.string.isRequired,
    duracion: PropTypes.number.isRequired,
    precio: PropTypes.number.isRequired,
  }),
};
