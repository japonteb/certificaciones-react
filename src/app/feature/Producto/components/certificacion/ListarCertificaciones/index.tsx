import * as PropTypes from 'prop-types';
import * as React from 'react';
import { ContainerParagraph, Table } from 'app/shared/components/Layout/styles';
import { BtnEliminarCertificacion } from '../EliminarCertificacion';
import { Certificacion } from '../../../models/Certificacion';

export interface ListaCertificacionesProps {
  certificaciones: Array<Certificacion>;
  onClickEliminarCertificacion: (certificaciones: Certificacion) => void;
}

export const ListaCertificaciones: React.FC<ListaCertificacionesProps> = ({
  certificaciones,
  onClickEliminarCertificacion,
}) => {
  return (
    <>
      <h2>Certificaciones</h2>
      <ContainerParagraph>Certificaciones disponibles.</ContainerParagraph>
      <Table>
        <thead>
          <tr>
            <td>
              <b>Nombre</b>
            </td>
            <td>
              <b>Detalle</b>
            </td>
            <td>
              <b>Duraci√≥n</b>
            </td>
            <td>
              <b>Precio</b>
            </td>
            <td>
              <b>Eliminar</b>
            </td>
          </tr>
        </thead>
        <tbody>
          {certificaciones.map((certificacion: Certificacion) => {
            return (
              <tr key={certificacion.id}>
                <td>{certificacion.nombre}</td>
                <td>{certificacion.detalle}</td>
                <td>{certificacion.duracion}</td>
                <td>{certificacion.precio}</td>
                <td>
                  <BtnEliminarCertificacion
                    certificacion={certificacion}
                    onEliminar={onClickEliminarCertificacion}
                  ></BtnEliminarCertificacion>
                </td>
              </tr>
            );
          })}
        </tbody>
      </Table>
    </>
  );
};

ListaCertificaciones.propTypes = {
  certificaciones: PropTypes.array.isRequired,
  onClickEliminarCertificacion: PropTypes.func.isRequired,
};
