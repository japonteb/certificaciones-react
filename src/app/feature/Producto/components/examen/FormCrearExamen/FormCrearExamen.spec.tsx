import * as React from 'react';
import { RenderResult, fireEvent, render, wait } from '@testing-library/react';
import { SinonStub, stub } from 'sinon';
import { FormCrearCertificacion } from '../../certificacion/FormCrearCertificacion';
import { setTextEvent } from 'app/shared/utils/test';

describe('FormCrearCertificacion test', () => {
  let componentWrapper: RenderResult;
  let componentProps: React.ComponentProps<typeof FormCrearCertificacion> & {
    onSubmit: SinonStub;
  };

  beforeEach(() => {
    componentProps = {
      formTitle: 'Form test',
      onSubmit: stub(),
    };
    componentWrapper = render(<FormCrearCertificacion {...componentProps} />);
  });

  it('should match snapshot', () => {
    expect(componentWrapper.container).toMatchSnapshot();
  });

  it('should fail on submit all fields missing', async () => {
    const elem = componentWrapper.container;
    const submitButton = elem.querySelector('button[type="submit"]');

    await wait(() => {
      submitButton && fireEvent.click(submitButton);
    });
    const spans = elem.querySelectorAll('span');
    expect(spans.length).toBe(4);
    expect(spans[0].textContent).toBe('El campo nombre es requerido.');
    expect(spans[1].textContent).toBe('El campo detalle es requerido.');
    expect(spans[2].textContent).toBe('El campo duración es requerido.');
    expect(spans[3].textContent).toBe('El campo precio es requerido.');
  });

  it('should fail on submit three fields missing', async () => {
    const elem = componentWrapper.container;
    const nombre = elem.querySelector('input[name="nombre"]');
    const submitButton = elem.querySelector('button[type="submit"]');

    await wait(() => {
      nombre && fireEvent.change(nombre, setTextEvent('nombre', 'Lorem'));
    });

    await wait(() => {
      submitButton && fireEvent.click(submitButton);
    });
    const spans = elem.querySelectorAll('span');
    expect(spans.length).toBe(3);
    expect(spans[0].textContent).toBe('El campo detalle es requerido.');
    expect(spans[1].textContent).toBe('El campo duración es requerido.');
    expect(spans[2].textContent).toBe('El campo precio es requerido.');
  });

  it('should fail on submit two fields missing', async () => {
    const elem = componentWrapper.container;

    const nombre = elem.querySelector('input[name="nombre"]');
    const detalle = elem.querySelector('input[name="detalle"]');
    const submitButton = elem.querySelector('button[type="submit"]');

    await wait(() => {
      nombre && fireEvent.change(nombre, setTextEvent('detalle', 'Lorem'));
    });
    await wait(() => {
      detalle && fireEvent.change(detalle, setTextEvent('detalle', 'Ipsum'));
    });

    await wait(() => {
      submitButton && fireEvent.click(submitButton);
    });
    const spans = elem.querySelectorAll('span');
    expect(spans.length).toBe(2);
    expect(spans[0].textContent).toBe('El campo duración es requerido.');
    expect(spans[1].textContent).toBe('El campo precio es requerido.');
  });

  it('should fail on submit one field missing', async () => {
    const elem = componentWrapper.container;

    const nombre = elem.querySelector('input[name="nombre"]');
    const detalle = elem.querySelector('input[name="detalle"]');
    const duracion = elem.querySelector('input[name="duracion"]');
    const submitButton = elem.querySelector('button[type="submit"]');

    await wait(() => {
      nombre && fireEvent.change(nombre, setTextEvent('detalle', 'Lorem'));
    });
    await wait(() => {
      detalle && fireEvent.change(detalle, setTextEvent('detalle', 'Ipsum'));
    });

    await wait(() => {
      duracion && fireEvent.change(duracion, setTextEvent('duracion', 'Ipsum'));
    });

    await wait(() => {
      submitButton && fireEvent.click(submitButton);
    });
    const spans = elem.querySelectorAll('span');
    expect(spans.length).toBe(1);
    expect(spans[1].textContent).toBe('El campo precio es requerido.');
  });

  it('should submit', async () => {
    const elem = componentWrapper.container;

    const nombre = elem.querySelector('input[name="nombre"]');
    const detalle = elem.querySelector('input[name="detalle"]');
    const duracion = elem.querySelector('input[name="duracion"]');
    const precio = elem.querySelector('input[name="precio"]');
    const submitButton = elem.querySelector('button[type="submit"]');

    await wait(() => {
      nombre && fireEvent.change(nombre, setTextEvent('nombre', 'Lorem'));
    });
    await wait(() => {
      detalle && fireEvent.change(detalle, setTextEvent('detalle', 'Ipsum'));
    });
    await wait(() => {
      duracion && fireEvent.change(duracion, setTextEvent('duracion', '120'));
    });
    await wait(() => {
      precio && fireEvent.change(precio, setTextEvent('precio', '1000'));
    });

    await wait(() => {
      submitButton && fireEvent.click(submitButton);
    });

    const formSubmitted = componentProps.onSubmit.firstCall.args[0];

    expect(formSubmitted.nombre).toBe('Lorem');
    expect(formSubmitted.detalle).toBe('Ipsum');
    expect(formSubmitted.duracion).toBe('120');
    expect(formSubmitted.precio).toBe('1000');
  });
});
