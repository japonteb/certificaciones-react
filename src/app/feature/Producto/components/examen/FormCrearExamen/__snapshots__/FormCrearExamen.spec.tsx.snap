// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormCrearExamen test should match snapshot 1`] = `
<div>
  <form>
    <h2>
      Form test
    </h2>
    <input
      class="sc-AxirZ eAukyj"
      name="nombre"
      placeholder="Nombre de la certificación"
      value=""
    />
    <input
      class="sc-AxirZ eAukyj"
      name="detalle"
      placeholder="Detalle de la certificación"
      value=""
    />
    <input
      class="sc-AxirZ eAukyj"
      name="duracion"
      placeholder="Duración en horas"
      value=""
    />
    <input
      class="sc-AxirZ eAukyj"
      name="precio"
      placeholder="Precio en COP"
      value=""
    />
    <button
      class="sc-AxjAm kgxTKy"
      type="submit"
    >
      Registrar
    </button>
  </form>
</div>
`;
